knodle/trainer/crossweigh_weighing/data_splitting_by_rules.py

```diff
@@ -0,0 +1,289 @@
+        seed: int = None, other_class_id: int = None
+) -> (List, List):
+    """
+    This function allows to perform the splitting of the data instances into k folds according to the signatures.
+    The sample signature is composed from the rule matched in the sample. For example, if rules with ids 1, 5, 7 matched
+    in the sample, the sample signature is 1_5_7. Thus, the signature serves as a sort of sample identification.
+    The logic is the following:
+        - the sample signatures are calculated
+        for each partition:
+        - the sample signatures are shuffled
+        - the sample signatures are splitted into k folds
+        - each fold iteratively becomes a hold-out fold
+        - the samples with signatures from the hold-out fold are added to the hold-out test set
+        - other samples are added to the train set
+    The train and test sets do not intersect.
+
+    :param data_features: encoded data samples (num_samples x num_features)
+    :param labels: array of labels (num_samples x 1)
+    :param rule_matches_z: matrix of rule matches (num_samples x num_rules)
+    :param partitions: number of partitions that are to be performed; in each partition the dataset will be splitted into
+    k folds
+    :param folds: number of folds the data instances are to be slitted into in each partition
+    :param seed: optionally, the seed could be fixed in order to provide reproducibility
+    :param other_class_id: if you don't want to include the negative samples (the ones that belong to the other class)
+     to the test set, but only to the training set, you can pass the id of other class should be
+
+    :return: two lists; the first contains the training sets, the second contains the test (hold-out) sets.
+    """
+
+    random.seed(seed) if seed is not None else random.choice(range(9999))
+    signature2id, signature_id2samples_ids = get_signature_sample_ids(rule_matches_z)
+
+    return compose_train_n_test_datasets(
+        data_features, signature_id2samples_ids, labels, folds, partitions, other_class_id
+    )
+
+
+def get_rules_sample_ids(rule_matches_z: Union[np.ndarray, sp.csr_matrix]) -> Dict:
+    """
+    This function creates a dictionary {rule id : sample id where this rule matched}. The dictionary is needed as a
+    support tool for faster calculation of train and test sets.
+    :param rule_matches_z: matrix of rule matches as a numpy or sparse matrix (num_samples x num_rules)
+    :return: dictionary {rule_id: [sample_id1, sample_id2, ...]}
+    """
+    if isinstance(rule_matches_z, sp.csr_matrix):
+        rules_samples_ids_dict = {key: set() for key in range(rule_matches_z.shape[1])}
+        for row_id, col_id in zip(*rule_matches_z.nonzero()):       # row_id = sample id, col_id = rule id
+            rules_samples_ids_dict[col_id].add(row_id)
+    else:
+        rules_samples_ids_dict = dict((i, set()) for i in range(0, rule_matches_z.shape[1]))
```

Comment:
same reason as in signatures
```suggestion
        rules_samples_ids_dict = {}
```